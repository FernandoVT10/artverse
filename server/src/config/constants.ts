import { config } from "dotenv";
// set all variables from ".env" file in the "process.env"
config();

type DBConfig = {
  host: string;
  username: string;
  password: string;
  databaseName: string;
  port: number;
};

const dbPort = parseInt(process.env.PORT || "") || 3306;

export const DB_CONFIG: DBConfig = {
  host: process.env.DB_HOST || "",
  username: process.env.DB_USERNAME || "",
  password: process.env.DB_PASSWORD || "",
  databaseName: process.env.DB_NAME || "",
  port: dbPort,
};
