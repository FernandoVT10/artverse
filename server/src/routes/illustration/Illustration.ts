import {
  Association,
  Model,
  InferAttributes,
  InferCreationAttributes,
  CreationOptional,
  DataTypes,
  ForeignKey,
  Sequelize,
} from "sequelize";

import { User } from "@routes/user/User";

// this type has the values that are neccesary to create the IllustrationImages model
export type IllustrationImagesType = {
  thumbnail: string;
  original: string;
};

class IllustrationImages extends Model<
  InferAttributes<IllustrationImages>,
  InferCreationAttributes<IllustrationImages>
> {
  declare id: CreationOptional<number>;
  declare illustrationId: ForeignKey<Illustration["id"]>;
  declare thumbnail: string;
  declare original: string;

  declare static associations: {
    illustration: Association<IllustrationImages, Illustration>;
  };
}

export class Illustration extends Model<
  InferAttributes<Illustration>,
  InferCreationAttributes<Illustration>
> {
  declare id: CreationOptional<number>;
  declare userId: ForeignKey<User["id"]>;
  declare title: string;
  declare description: string;

  declare createdAt: CreationOptional<Date>;
  declare updatedAt: CreationOptional<Date>;

  declare static associations: {
    images: Association<Illustration, IllustrationImages>;
    user: Association<Illustration, User>;
  };
}

export default {
  init(sequelize: Sequelize) {
    Illustration.init(
      {
        id: {
          type: DataTypes.INTEGER.UNSIGNED,
          autoIncrement: true,
          primaryKey: true,
        },
        title: {
          type: DataTypes.STRING(100),
          allowNull: false,
        },
        description: DataTypes.STRING(500),
        createdAt: DataTypes.DATE,
        updatedAt: DataTypes.DATE,
      },
      {
        sequelize,
        tableName: "illustrations",
      }
    );

    IllustrationImages.init(
      {
        id: {
          type: DataTypes.INTEGER.UNSIGNED,
          autoIncrement: true,
          primaryKey: true,
        },
        thumbnail: DataTypes.STRING,
        original: DataTypes.STRING,
      },
      {
        sequelize,
        tableName: "illustrations_images",
      }
    );

    return Illustration;
  },

  associate(models: Sequelize["models"]) {
    Illustration.belongsTo(models.User, {
      as: "user",
    });

    Illustration.hasOne(IllustrationImages, {
      as: "images",
      foreignKey: {
        name: "illustrationId",
        allowNull: false,
      },
      onDelete: "CASCADE",
    });

    IllustrationImages.belongsTo(Illustration, {
      as: "illustration",
    });
  },
};

// export default Illustration;
