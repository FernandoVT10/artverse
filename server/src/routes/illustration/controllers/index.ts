export * as createIllustration from "./createIllustration";
