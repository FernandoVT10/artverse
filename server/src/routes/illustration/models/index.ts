import IllustrationImagesFactory from "./IllustrationImages";
import IllustrationFactory from "./Illustration";
import LikeFactory from "./Like";

export { IllustrationImages } from "./IllustrationImages";
export { Illustration } from "./Illustration";
export { Like } from "./Like";

export default [IllustrationImagesFactory, IllustrationFactory, LikeFactory];
