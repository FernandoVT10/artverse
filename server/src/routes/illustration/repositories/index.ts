export { default as createIllustration } from "./createIllustration";
