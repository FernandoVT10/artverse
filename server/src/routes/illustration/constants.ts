import path from "path";
import { IMAGES_DIR } from "@config/constants";
import { Size } from "@utils/images/saveAndResizeMulterFileAsImage";

export const ILLUSTRATIONS_DESTINATION = path.resolve(
  IMAGES_DIR,
  "./illustrations/"
);

export const ILLUSTRATIONS_SIZES: Size[] = [
  {
    width: 250,
    height: 250,
    suffix: "thumbnail",
  },
];
