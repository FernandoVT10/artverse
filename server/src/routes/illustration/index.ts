import { Router } from "express";
import { createIllustration } from "./controllers";

const router = Router();

router.post(
  "/illustrations/",
  createIllustration.middlewares(),
  createIllustration.controller
);

export default router;
