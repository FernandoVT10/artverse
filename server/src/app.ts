import express from "express";
import bodyParser from "body-parser";
import routes from "./routes";
import errorHandler from "@middlewares/errorHandler";

import { PUBLIC_DIR } from "./config/constants";

const app = express();

app.use(express.static(PUBLIC_DIR));
app.use(bodyParser.json());

app.use(routes);
app.use(errorHandler());

export default app;
