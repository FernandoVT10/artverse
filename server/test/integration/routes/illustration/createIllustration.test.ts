import fs from "fs";
import sharp from "sharp";
import path from "path";
import app from "@app";
import createAuthorizationTests from "../shared/createAuthorizationTests";

import supertest, { Response } from "supertest";

import { ILLUSTRATIONS_DESTINATION } from "@routes/illustration/constants";
import { generateJWT } from "@test-utils/factories/userFactory";
import { Illustration } from "@models";
import { clearDB } from "@test-utils/setupDB";

import {
  createDescriptionValidationTests,
  createImageValidationTests,
  createTitleValidationTests,
} from "./createValidationTests";

describe("Integration POST /api/illustrations/", () => {
  const request = supertest(app);
  const requestApi = () => request.post("/api/illustrations");

  const testImagePath = path.resolve(__dirname, "../../../fixtures/image.png");

  const illustrationMock = {
    title: "bar",
    description: "foo",
  };

  describe("authorization", () => {
    createAuthorizationTests(requestApi);
  });

  describe("validation", () => {
    let token: string;

    beforeAll(async () => {
      await clearDB();
      const result = await generateJWT();
      token = result.token;
    });

    describe("title", () => {
      const localRequestApi = () =>
        requestApi()
          .field("description", illustrationMock.description)
          .attach("image", testImagePath)
          .set("Cookie", `token=${token}`);

      createTitleValidationTests(localRequestApi);
    });

    describe("description", () => {
      const localRequestApi = () =>
        requestApi()
          .field("title", illustrationMock.title)
          .attach("image", testImagePath)
          .set("Cookie", `token=${token}`);

      createDescriptionValidationTests(localRequestApi);
    });

    describe("image", () => {
      const localRequestApi = () =>
        requestApi().field(illustrationMock).set("Cookie", `token=${token}`);

      createImageValidationTests(localRequestApi);
    });
  });

  describe("when the request is a success", () => {
    let response: Response;
    let userId: number;

    beforeAll(async () => {
      await clearDB();

      const { token, user } = await generateJWT();
      userId = user.id;

      response = await requestApi()
        .field(illustrationMock)
        .set("Cookie", `token=${token}`)
        .attach("image", testImagePath);
    });

    afterAll(async () => {
      const { original, thumbnail } = response.body.images;

      const originalName = path.basename(original);
      const thumbnailName = path.basename(thumbnail);

      // cleanup the uploaded images
      // TODO we need to change it for a better tested function in the future
      await Promise.all([
        fs.promises.rm(path.resolve(ILLUSTRATIONS_DESTINATION, originalName)),
        fs.promises.rm(path.resolve(ILLUSTRATIONS_DESTINATION, thumbnailName)),
      ]);
    });

    const expectedIllustration = {
      title: illustrationMock.title,
      description: illustrationMock.description,
      images: {
        original: expect.any(String),
        thumbnail: expect.any(String),
      },
    };

    it("should return a successfully response", async () => {
      expect(response.statusCode).toBe(200);
      expect(response.body).toMatchObject({
        ...expectedIllustration,
        userId,
      });
    });

    it("should save the illustration on the database", async () => {
      const illustration = await Illustration.findOne({
        include: Illustration.associations.images,
      });

      expect(illustration).toMatchObject({
        ...expectedIllustration,
        userId,
      });
    });

    const getMetadataFromURL = (url: string) => {
      const basename = path.basename(url);
      const fullPath = path.resolve(ILLUSTRATIONS_DESTINATION, basename);

      return sharp(fullPath).metadata();
    };

    it("should save the original image", async () => {
      const originalImageMetadata = await sharp(testImagePath).metadata();
      const uploadedImageMetadata = await getMetadataFromURL(
        response.body.images.original
      );

      expect(originalImageMetadata).toEqual(uploadedImageMetadata);
    });

    it("should save a thumbnail", async () => {
      expect(
        await getMetadataFromURL(response.body.images.thumbnail)
      ).toMatchObject({
        width: 250,
        height: 250,
        format: "webp",
      });
    });
  });
});
