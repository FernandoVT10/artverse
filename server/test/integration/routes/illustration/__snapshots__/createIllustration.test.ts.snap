// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration POST /api/illustrations/ authorization should return an error when the token is invalid 1`] = `
Array [
  Object {
    "message": "The credentials are invalid",
  },
]
`;

exports[`Integration POST /api/illustrations/ authorization should return an error when the userId of the token doesn't exist 1`] = `
Array [
  Object {
    "message": "There was an error trying to process your request. Please try again later.",
  },
]
`;

exports[`Integration POST /api/illustrations/ authorization should return an error when you don't create a cookie with the token 1`] = `
Array [
  Object {
    "message": "You don't have enough permissions",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation description should return an error when the description has more than 500 characters 1`] = `
Array [
  Object {
    "field": "description",
    "message": "The description must have 500 or less characters",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation description should return an error when the description is not a string 1`] = `
Array [
  Object {
    "field": "description",
    "message": "The description must be a string",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation description should return an error when the description is not provided 1`] = `
Array [
  Object {
    "field": "description",
    "message": "The description is required",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation image should return an error when the image is not attached 1`] = `
Array [
  Object {
    "field": "image",
    "message": "The image is required",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation image should return an error when you attach a file that's not an image 1`] = `
Array [
  Object {
    "field": "image",
    "message": "The field image must be an image",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation title should return an error when the title has more than 100 characters 1`] = `
Array [
  Object {
    "field": "title",
    "message": "The title must have 100 or less characters",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation title should return an error when the title is not a string 1`] = `
Array [
  Object {
    "field": "title",
    "message": "The title must be a string",
  },
]
`;

exports[`Integration POST /api/illustrations/ validation title should return an error when the title is not provided 1`] = `
Array [
  Object {
    "field": "title",
    "message": "The title is required",
  },
]
`;
